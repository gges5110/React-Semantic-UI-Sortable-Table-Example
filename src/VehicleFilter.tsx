import React from "react";
import PropTypes from "prop-types";
import { Form, Popup } from "semantic-ui-react";

const regex = new RegExp("^[a-zA-Z0-9 ]+$");

interface VehicleFilterProps {
  totalCount: number;
  loading?: boolean;
  onSubmitFilter(value: string): void;
}

interface VehicleFilterState {
  [index: string]: any;
  filter: string;
  filterValid: boolean;
}

export class VehicleFilter extends React.Component<
  VehicleFilterProps,
  VehicleFilterState
> {
  static propTypes = {
    onSubmitFilter: PropTypes.func.isRequired,
    filter: PropTypes.string.isRequired,
    totalCount: PropTypes.number.isRequired
  };

  constructor(props: VehicleFilterProps) {
    super(props);
    this.state = {
      filter: "",
      filterValid: true
    };
  }

  handleOnChange = (event: any, { name, value }: any) => {
    if (value !== "" && !regex.test(value)) {
      this.setState({ [name]: value, filterValid: false });
    } else {
      this.setState({ [name]: value, filterValid: true });
      this.props.onSubmitFilter(value);
    }
  };

  render() {
    const { filter } = this.state;
    let popupMessage = "";
    if (!this.state.filterValid) {
      popupMessage = "Invalid character.";
    } else if (this.props.totalCount === 0) {
      popupMessage = "No results found.";
    }

    return (
      <Form>
        <Form.Group>
          <Form.Field>
            <Popup
              trigger={
                <Form.Input
                  placeholder="Enter the filter."
                  name="filter"
                  value={filter}
                  error={!this.state.filterValid}
                  label="Filter"
                  onChange={this.handleOnChange}
                  icon="search"
                  loading={this.props.loading}
                />
              }
              content={popupMessage}
              on="click"
              open={!this.state.filterValid || this.props.totalCount === 0}
              position="right center"
            />
          </Form.Field>
        </Form.Group>
      </Form>
    );
  }
}
