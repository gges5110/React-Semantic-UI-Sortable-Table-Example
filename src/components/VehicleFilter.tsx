import React, { useState } from "react";
import { Form, Popup } from "semantic-ui-react";
import { InputOnChangeData } from "semantic-ui-react/dist/commonjs/elements/Input/Input";
import debounce from "lodash.debounce";

const regex = new RegExp("^[a-zA-Z0-9 ]+$");

interface VehicleFilterProps {
  filter: string;
  totalCount: number;
  loading?: boolean;
  onSubmitFilter(value: string): void;
}

interface VehicleFilterState {
  filter: string;
  filterValid: boolean;
}

export const VehicleFilter: React.FC<VehicleFilterProps> = ({
  totalCount,
  loading,
  onSubmitFilter,
  filter,
}) => {
  const [state, setState] = useState<VehicleFilterState>({
    filter: filter,
    filterValid: true,
  });

  const f = debounce((value: string) => {
    if (value !== "" && !regex.test(value)) {
      setState({ filter: value, filterValid: false });
    } else {
      setState({ filter: value, filterValid: true });
      onSubmitFilter(value);
    }
  }, 500);

  const handleOnChange = (
    event: React.ChangeEvent<HTMLInputElement>,
    { value }: InputOnChangeData
  ) => {
    f(value);
  };

  let popupMessage = "";
  if (!state.filterValid) {
    popupMessage = "Invalid character.";
  } else if (totalCount === 0) {
    popupMessage = "No results found.";
  }

  return (
    <Form>
      <Form.Group>
        <Form.Field>
          <Popup
            trigger={
              <Form.Input
                placeholder={"Enter a filter."}
                name={"filter"}
                error={!state.filterValid}
                label={"Filter"}
                onChange={handleOnChange}
                icon={"search"}
                loading={loading}
              />
            }
            content={popupMessage}
            on={"click"}
            open={!state.filterValid || totalCount === 0}
            position={"right center"}
          />
        </Form.Field>
      </Form.Group>
    </Form>
  );
};
