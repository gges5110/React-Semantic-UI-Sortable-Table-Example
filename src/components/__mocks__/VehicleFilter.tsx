import React from "react";

export const VehicleFilter: React.FC = () => <div>VehicleFilter</div>;
