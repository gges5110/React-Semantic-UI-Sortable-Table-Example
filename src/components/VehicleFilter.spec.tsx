import React from "react";
import { VehicleFilter } from "./VehicleFilter";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

jest.mock("lodash.debounce", () => jest.fn(fn => fn));

describe("VehicleFilter", () => {
  it("renders", () => {
    const filter = "";
    const totalCount = 100;
    const onSubmitFilter = jest.fn();

    const { container } = render(
      <VehicleFilter
        filter={filter}
        totalCount={totalCount}
        onSubmitFilter={onSubmitFilter}
      />
    );

    expect(container).toMatchSnapshot();
    expect(onSubmitFilter).not.toHaveBeenCalled();
  });

  it("should show message when 0 total count", () => {
    const filter = "";
    const totalCount = 0;
    const onSubmitFilter = jest.fn();

    render(
      <VehicleFilter
        filter={filter}
        totalCount={totalCount}
        onSubmitFilter={onSubmitFilter}
      />
    );

    waitFor(() => expect(screen.getByText("results")).toBeInTheDocument());
  });

  it("submits", () => {
    const filter = "";
    const totalCount = 10;
    const onSubmitFilter = jest.fn();

    render(
      <VehicleFilter
        filter={filter}
        totalCount={totalCount}
        onSubmitFilter={onSubmitFilter}
      />
    );

    expect(onSubmitFilter).not.toHaveBeenCalled();

    userEvent.type(screen.getByRole("textbox"), "Volvo");

    expect(onSubmitFilter).toHaveBeenCalled();
    expect(onSubmitFilter.mock.calls).toHaveLength(5);
  });

  it("submits invalid", () => {
    const filter = "";
    const totalCount = 10;
    const onSubmitFilter = jest.fn();

    render(
      <VehicleFilter
        filter={filter}
        totalCount={totalCount}
        onSubmitFilter={onSubmitFilter}
      />
    );

    userEvent.type(screen.getByRole("textbox"), "#");

    expect(onSubmitFilter).not.toHaveBeenCalled();
    waitFor(() => expect(screen.getByText("Invalid")).toBeInTheDocument());
  });
});
