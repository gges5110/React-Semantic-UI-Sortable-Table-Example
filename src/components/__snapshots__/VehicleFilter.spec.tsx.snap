// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleFilter renders 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="fields"
    >
      <div
        class="field"
      >
        <div
          class="field"
        >
          <label>
            Filter
          </label>
          <div
            class="ui icon input"
          >
            <input
              name="filter"
              placeholder="Enter a filter."
              type="text"
            />
            <i
              aria-hidden="true"
              class="search icon"
            />
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
